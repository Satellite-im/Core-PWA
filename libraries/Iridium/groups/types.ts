import { IridiumDocument } from '@satellite-im/iridium'
import Group from './Group'

export type GroupMemberDetails = {
  id: string
  name: string
  photoHash: string
}

export type GroupConfig = {
  name: string
  origin: string
  avatar?: string
  members: { [did: string]: GroupMemberDetails }
  metadata?: IridiumDocument
}

export type GroupData = GroupConfig & {
  id: string
}

export type GroupMap = { [key: string]: Group }

export type GroupManagerEvent =
  | 'group-created'
  | 'group-updated'
  | 'group-deleted'
  | 'group-member-added'
  | 'group-member-removed'
  | 'group-member-updated'
  | 'group-member-joined'
  | 'group-member-left'

export enum GroupsError {
  NETWORK_ERROR = 'errors.groups.network',
  NOT_A_MEMBER = 'errors.groups.not_a_member',
  GROUP_ALREADY_EXISTS = 'error.groups.group_already_exists',
  GROUP_NOT_FOUND = 'error.groups.group_not_found',
  GROUP_NOT_INITIALIZED = 'errors.groups.group_not_initialized',
  INVITE_ALREADY_SENT = 'errors.groups.request_already_sent',
  RECIPIENT_NOT_FOUND = 'errors.groups.recipient_not_found',
  USER_NOT_INITIALIZED = 'errors.groups.user_not_initialized',
}
