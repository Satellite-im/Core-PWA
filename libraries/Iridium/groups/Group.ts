import { Emitter } from '@satellite-im/iridium'
import type {
  IridiumMessage,
  IridiumChannelEvent,
  IridiumDocument,
  IridiumGetOptions,
  IridiumPayload,
  IridiumPeerMessage,
  IridiumSetOptions,
} from '@satellite-im/iridium/src/types'
import { ConversationMessage } from '~/libraries/Iridium/chat/types'
import { IridiumManager } from '~/libraries/Iridium/IridiumManager'
import { GroupData, GroupsError } from '~/libraries/Iridium/groups/types'

export default class Group extends Emitter<IridiumMessage> {
  constructor(
    public readonly id: string,
    private readonly iridium: IridiumManager,
    private state?: GroupData,
  ) {
    super()
  }

  get name() {
    return this.state?.name
  }

  get members() {
    return this.state?.members
  }

  static async getById(id: string, iridium: IridiumManager) {
    const group = new Group(id, iridium)
    await group.load()
    return group
  }

  async load() {
    if (!this.iridium.connector) {
      throw new Error(GroupsError.GROUP_NOT_INITIALIZED)
    }

    this.state = await this.iridium.connector?.get(`/groups/${this.id}`)
    this.iridium.connector.on(`/groups/${this.id}`, this._onWireMessage)
    await this.iridium.connector.subscribe(`/groups/${this.id}`)
  }

  private _onWireMessage(message: IridiumPeerMessage<IridiumChannelEvent>) {
    const { from, payload } = message
    const { channel, data } = payload
    this.emit(channel, { from, data, payload })
  }

  assertReady() {
    if (!this.state?.members) {
      throw new Error(GroupsError.GROUP_NOT_INITIALIZED)
    }
  }

  sendChatMessage(message: ConversationMessage) {
    return this.iridium.chat?.sendMessage(this.id, message)
  }

  send(payload: ConversationMessage) {
    if (!this.state?.members) {
      throw new Error(GroupsError.GROUP_NOT_INITIALIZED)
    }

    console.info('sending group message')
    return this.iridium.connector?.send(payload, {
      to: this.state?.members,
    })
  }

  broadcast<T = IridiumDocument | string>(channel: string, data: T) {
    this.assertReady()
    return this.iridium.connector?.broadcast(
      `/groups/${this.id}`,
      { channel, data },
      { encrypt: { recipients: this.state?.members } },
    )
  }

  async set(
    key: string,
    payload: IridiumPayload,
    options: IridiumSetOptions = {},
  ) {
    this.assertReady()

    if (!options.store) {
      options.store = {}
    }
    if (!options.store.encrypt) {
      options.store.encrypt = {}
    }
    options.store.encrypt.recipients = this.state?.members // TODO: to be fixed
    return this.iridium.connector?.set(
      `/groups/${this.id}/${key}`,
      payload,
      options,
    )
  }

  async get(key: string, options?: IridiumGetOptions) {
    this.assertReady()
    return this.iridium.connector?.get(`/groups/${this.id}/${key}`, options)
  }
}
