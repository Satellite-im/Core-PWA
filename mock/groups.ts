import { Group } from '~/types/ui/core'

export const Groups: Group[] = [
  {
    name: 'Satellite',
    address: '0x9bf4001d307dfd62b26a2f1307ee0c0307632d59',
    motd: 'Chatting privatley with the world',
    members: [
      '0x0ED3B01FAF2002beD1434Fce6E47aDa3bD43d213',
      '0xca8d2554190310F0DdAcF8647b700ceAABc4Cb10',
    ],
  },
  {
    name: 'Solstice',
    address: '0x07ee55aa48bb72dcc6e9d78256648910de513eca',
    motd: "Chillin'n in the sun",
    members: [
      '0x9bf4001d307dfd62b26a2f1307ee0c0307632d59',
      '0x0ED3B01FAF2002beD1434Fce6E47aDa3bD43d213',
      '0x7cFbE9977E6577117B589e15CeF2f2F6d18F64ff',
      '0xca8d2554190310F0DdAcF8647b700ceAABc4Cb10',
    ],
  },
  {
    name: 'Urania',
    address: '0x07ee55aa48bb72dcc6e9d71256648910de513eca',
    motd: 'Amusing universe',
    members: ['0xb2D822e05176563498901682202A1f9c0F68C93c'],
  },
]

export const ExampleGroup: Group = {
  name: 'Solstice',
  address: '0x9bf4001d307dfd62b26a2f1307ee0c0307632d59',
  motd: "Chillin'n in the sun",
  members: [],
}
