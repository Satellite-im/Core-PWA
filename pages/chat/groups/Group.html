<div id="group">
  <UiChatdEncrypted />
  <div v-on:click="sendMessage">send message</div>
  <Conversation :messages="groupMessages.message" :loading="false" />
</div>
