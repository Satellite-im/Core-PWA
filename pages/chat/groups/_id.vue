<template src="./Group.html" />

<script lang="ts">
import Vue from 'vue'
import { mapState } from 'vuex'
import { groupMessages } from '~/utilities/Messaging'

export default Vue.extend({
  name: 'GroupMessages',
  layout: 'chat',
  data() {
    return {
      groupMessages: [],
      groupID: '',
      groupMembers: [],
    }
  },
  head() {
    return {
      title: this.meta.title,
    }
  },
  computed: {
    ...mapState(['media', 'meta']),
    groupedMessages() {
      const { id } = this.$route.params
      const conversation = this.$typedStore.state.textile.conversations[id]

      if (!conversation) {
        return []
      }

      const { messages, replies, reactions } = conversation

      return groupMessages(messages, replies, reactions)
    },
  },
  mounted() {
    const { id } = this.$route.params
    this.getMessages(id)
    this.subscribeToGroup(id)
    this.fetchGroupMembers(id)
  },
  methods: {
    async getMessages(id: string) {
      this.$data.groupID = id
      this.$data.groupMessages = await this.$store.dispatch(
        'textile/fetchGroupMessages',
        { groupId: id },
      )
    },
    async subscribeToGroup(id: string) {
      this.$data.groupMessages = await this.$store.dispatch(
        'textile/subscribeToGroup',
        { groupId: id },
      )
    },
    async fetchGroupMembers(id: string) {
      await this.$store.dispatch('groups/fetchGroupMembers', id)
    },
  },
})
</script>

<style lang="less" src="./Group.less"></style>
