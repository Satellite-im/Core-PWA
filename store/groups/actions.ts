import Vue from 'vue'
import { map, uniq } from 'lodash'
import { DataStateType } from '../dataState/types'
import { Group, GroupMember, GroupsError, GroupsState } from './types'
import SolanaManager from '~/libraries/Solana/SolanaManager/SolanaManager'
import { ActionsArguments } from '~/types/store/store'
import TextileManager from '~/libraries/Textile/TextileManager'
import { GroupChatManager } from '~/libraries/Textile/GroupChatManager'
import BlockchainClient from '~/libraries/BlockchainClient'

const getGroupChatManager = (): GroupChatManager => {
  const $TextileManager: TextileManager = Vue.prototype.$TextileManager

  if (!$TextileManager.groupChatManager?.isInitialized()) {
    throw new Error(GroupsError.TEXTILE_NOT_INITIALIZED)
  }
  return $TextileManager.groupChatManager
}

const getUserAccount = () => {
  try {
    return BlockchainClient.getInstance().account
  } catch (e) {
    throw new Error(GroupsError.USER_NOT_INITIALIZED)
  }
}

export default {
  async initialize({ dispatch, state }: ActionsArguments<GroupsState>) {
    await Promise.all([
      dispatch('fetchGroups'),
      dispatch('subscribeToGroupInvites'),
      dispatch('subscribeToGroupsUpdate'),
    ])
    await Promise.all(
      state.all.map(async (group) => {
        await dispatch(
          'textile/subscribeToGroup',
          { groupId: group.id },
          { root: true },
        )
      }),
    )
  },

  /**
   * @method createGroup
   * @description create new group chat
   * @example
   */
  async createGroup(
    { commit }: ActionsArguments<GroupsState>,
    { name }: { name: string },
  ) {
    const $BlockchainClient: BlockchainClient = BlockchainClient.getInstance()
    const groupChatManager = getGroupChatManager()

    const groupId = await groupChatManager.createGroupConversation()

    await $BlockchainClient.createGroup(groupId, name)

    return groupId
  },

  /**
   * @method fetchGroups
   * @description fetch groups in which user participates
   * @example
   */
  async fetchGroups({ commit, dispatch }: ActionsArguments<GroupsState>) {
    commit(
      'dataState/setDataState',
      { key: 'groups', value: DataStateType.Loading },
      { root: true },
    )

    const $BlockchainClient: BlockchainClient = BlockchainClient.getInstance()
    const { address } = getUserAccount()

    const groups = await $BlockchainClient.getUserGroups(address)
    const groupsUsers = await $BlockchainClient.getGroupsUsers(
      map(groups, 'id'),
    )

    const addresses = uniq([].concat(...(<any>map(groupsUsers, 'users'))))
    const usersInfo = await $BlockchainClient.getUsersInfo(addresses)

    const payload = groups.map((group) => {
      const members = groupsUsers
        .find((it) => it.id === group.id)
        ?.users?.map((address) =>
          usersInfo.find((user) => user.address === address),
        )
        .filter(Boolean) as GroupMember[]

      return { ...group, members: members || [] }
    })
    commit('setGroups', payload)
    await dispatch('fetchGroupsLastUpdate')
    commit(
      'dataState/setDataState',
      { key: 'groups', value: DataStateType.Ready },
      { root: true },
    )
  },

  /**
   * @method sendGroupInvite
   * @description invites user into group chat
   * @param group {Group} target group
   * @param recipient {string} recipient address
   * @example
   */
  async sendGroupInvite(
    { dispatch }: ActionsArguments<GroupsState>,
    { group, recipient }: { group: Group; recipient: string },
  ) {
    const $BlockchainClient: BlockchainClient = BlockchainClient.getInstance()
    await $BlockchainClient.inviteToGroup(group.id, recipient)

    dispatch(
      'textile/sendGroupMessage',
      {
        groupId: group.id,
        message: `\`I added ${recipient} to the chat\``,
      },
      { root: true },
    )
  },

  /**
   * @method subscribeToGroupInvites
   * @description subscribe to group chat invites
   * @example
   */
  async subscribeToGroupInvites({
    state,
    dispatch,
    commit,
  }: ActionsArguments<GroupsState>) {
    if (state.inviteSubscription !== null) {
      await dispatch('unsubscribeFromGroupInvites')
    }

    const $BlockchainClient: BlockchainClient = BlockchainClient.getInstance()
    const id = $BlockchainClient.addGroupInviteListener((payload) => {
      commit('addGroup', { ...payload, members: [] })
      dispatch('fetchGroupMembers', payload.id)
      dispatch('subscribeToGroupUpdate', payload.id)
    })

    commit('setInviteSubscription', id)
  },

  /**
   * @method unsubscribeFromGroupInvites
   * @description remove group chat invites listener
   * @example
   */
  async unsubscribeFromGroupInvites({
    state,
    commit,
  }: ActionsArguments<GroupsState>) {
    if (state.inviteSubscription) {
      const $BlockchainClient: BlockchainClient = BlockchainClient.getInstance()
      await $BlockchainClient.unsubscribeGroupInvite(state.inviteSubscription)
      commit('setInviteSubscription', null)
    }
  },

  /**
   * @method subscribeToGroupsUpdate
   * @description subscribe to multiple groups update
   * @example
   */
  async subscribeToGroupsUpdate({
    state,
    dispatch,
    commit,
  }: ActionsArguments<GroupsState>) {
    if (state.groupSubscriptions.length) {
      await dispatch('unsubscribeFromGroupsUpdate')
    }
    const $BlockchainClient: BlockchainClient = BlockchainClient.getInstance()
    const ids = await $BlockchainClient.addGroupsListener((payload) => {
      commit('updateGroup', payload)
      dispatch('fetchGroupMembers', payload.id)
    })

    commit('setGroupSubscriptions', ids)
  },

  /**
   * @method subscribeToGroupUpdate
   * @description subscribe to group update
   * @example
   */
  async subscribeToGroupUpdate(
    { state, dispatch, commit }: ActionsArguments<GroupsState>,
    groupId: string,
  ) {
    await dispatch('textile/subscribeToGroup', { groupId }, { root: true })
    if (!state.groupSubscriptions.includes(groupId)) {
      const $BlockchainClient: BlockchainClient = BlockchainClient.getInstance()
      const id = await $BlockchainClient.addGroupListener(
        groupId,
        (payload) => {
          commit('updateGroup', payload)
          dispatch('fetchGroupMembers', payload.id)
        },
      )
      commit('addGroupSubscription', id)
    }
  },

  /**
   * @method unsubscribeFromGroupsUpdate
   * @description remove all groups update invites listener
   * @example
   */
  async unsubscribeFromGroupsUpdate({
    state,
    commit,
  }: ActionsArguments<GroupsState>) {
    const $BlockchainClient: BlockchainClient = BlockchainClient.getInstance()
    if (state.groupSubscriptions.length) {
      await $BlockchainClient.removeGroupListeners(state.groupSubscriptions)
      commit('setGroupSubscriptions', [])
    }
  },

  /**
   * @method addGroup
   * @description fetch group data and add it to group list
   * @param groupId {string} group id to add
   * @example
   */
  async addGroup(
    { commit, state }: ActionsArguments<GroupsState>,
    groupId: string,
  ) {
    if (!state.all.find((group) => group.id === groupId)) {
      const $BlockchainClient: BlockchainClient = BlockchainClient.getInstance()
      const group = await $BlockchainClient.getGroupById(groupId)

      commit('addGroup', group)
    }
  },

  /**
   * @method fetchGroupMembers
   * @description fetch groups members by given group id
   * @example
   */
  async fetchGroupMembers(
    { commit }: ActionsArguments<GroupsState>,
    groupId: string,
  ) {
    const $BlockchainClient: BlockchainClient = BlockchainClient.getInstance()

    const addresses = await $BlockchainClient.getGroupUsers(groupId)
    const members = await $BlockchainClient.getUsersInfo(addresses)
    commit('setGroupMembers', { groupId, members })
  },

  /**
   * @method fetchGroupsLastUpdate
   * @description fetch groups last message time
   * @example
   */
  async fetchGroupsLastUpdate({
    commit,
    state,
  }: ActionsArguments<GroupsState>) {
    const groupChatManager = getGroupChatManager()

    const ids = map(state.all, 'id')
    const results = await Promise.all(
      ids.map((id) => groupChatManager.getGroupLastUpdate(id)),
    )
    const payload = ids.reduce(
      (prev, curr, i) => ({ ...prev, [curr]: results[i] }),
      {},
    )
    commit('setGroupsLastUpdate', payload)
  },
}

export const exportForTesting = {
  getGroupChatManager,
  getUserAccount,
}
