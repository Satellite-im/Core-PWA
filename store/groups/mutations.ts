import { GroupsState, Group, GroupMember } from './types'

const mutations = {
  setGroups(state: GroupsState, groups: Group[]) {
    state.all = groups
  },
  addGroup(state: GroupsState, group: Group) {
    state.all.push(group)
  },
  updateGroup(state: GroupsState, group: Group) {
    state.all = state.all.map((item) =>
      item.id !== group.id ? item : { ...item, ...group },
    )
  },
  setInviteSubscription(state: GroupsState, value: number | null) {
    state.inviteSubscription = value
  },
  setGroupSubscriptions(state: GroupsState, value: string[]) {
    state.groupSubscriptions = value
  },
  addGroupSubscription(state: GroupsState, value: string) {
    state.groupSubscriptions.push(value)
  },
  setGroupMembers(
    state: GroupsState,
    { groupId, members }: { groupId: string; members: GroupMember[] },
  ) {
    const group = state.all.find((g) => g.id === groupId)
    if (group) {
      group.members = members.sort((a, b) => a.name.localeCompare(b.name))
    }
  },
  setGroupsLastUpdate(state: GroupsState, payload: { [key: string]: number }) {
    state.all = state.all.map((group) => ({
      ...group,
      lastUpdate: payload[group.id],
    }))
  },
}

export default mutations
