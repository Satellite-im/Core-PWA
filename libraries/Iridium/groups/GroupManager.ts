import { Emitter, encoding } from '@satellite-im/iridium'
import type { IridiumMessage } from '@satellite-im/iridium/src/types'
import { IridiumManager } from '../IridiumManager'
import Group from './Group'
import { GroupConfig, GroupsError } from './types'

export default class GroupManager extends Emitter<IridiumMessage> {
  groupIds?: string[]
  state: { [key: string]: Group } = {}

  constructor(private readonly iridium: IridiumManager) {
    super()
    this.iridium = iridium
  }

  async init() {
    const iridium = this.iridium.connector
    if (!iridium) {
      throw new Error('cannot initialize groups, no iridium connector')
    }

    await this.fetch()
  }

  private async fetch() {
    this.state = await this.iridium.connector?.get('/groups')
  }

  /**
   * @method createGroup
   * attempt to create conversation with the provided config. if successful, create a new group too
   * @returns a string UUID of the created groupChat
   */
  async createGroup(config: Omit<GroupConfig, 'origin'>): Promise<string> {
    if (!this.iridium.connector) {
      return ''
    }
    const id = await encoding.hash({
      timestamp: Date.now(),
      origin: this.iridium.connector?.id,
    })

    await this.iridium.chat.createConversation({
      id,
      name: config.name,
      type: 'group',
      participants: Object.values(config.members).map((m) => m.id),
    })
    await this.iridium.connector.set(`/groups/${id}`, {
      id,
      origin: this.iridium.connector.id,
      ...config,
    })
    return id
  }

  async getGroup(groupId: string) {
    if (this.state[groupId]) {
      return this.state[groupId]
    }

    const group = new Group(groupId, this.iridium)
    if (!group) {
      throw new Error(GroupsError.GROUP_NOT_FOUND)
    }
    return group
  }

  async getGroupMembers(groupId: string) {
    const group = await this.getGroup(groupId)
    return group.members
  }
}
