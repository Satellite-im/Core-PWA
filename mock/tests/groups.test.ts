import { expect } from '@jest/globals'
import * as Groups from '../groups'

describe('init', () => {
  it('gets the constant', () => {
    expect(Groups.Groups).toMatchSnapshot()
  })
})
