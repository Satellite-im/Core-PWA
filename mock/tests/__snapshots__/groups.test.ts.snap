// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`init gets the constant 1`] = `
Array [
  Object {
    "address": "af2e998e-bc77-4fca-ac0d-561078488c93",
    "creator": "0xdf9eb223bafbe5c5271415c75aecd68c21fe3d7f",
    "members": Array [
      "0xdf9eb223bafbe5c5271415c75aecd68c21fe3d7f",
      "0x07ee55aa48bb72dcc6e9d78256648910de513eca",
    ],
    "motd": "Chatting privately with the world",
    "name": "Satellite",
  },
  Object {
    "address": "bd30dd29-402a-46de-90f4-3ef78f4d0f7c",
    "creator": "0x07ee55aa48bb72dcc6e9d78256648910de513eca",
    "members": Array [
      "0xdf9eb223bafbe5c5271415c75aecd68c21fe3d7f",
      "0x07ee55aa48bb72dcc6e9d78256648910de513eca",
      "0x9bf4001d307dfd62b26a2f1307ee0c0307632d59",
      "0xc61b9bb3a7a0767e3179713f3a5c7a9aedce193c",
    ],
    "motd": "Chillin' in the sun",
    "name": "Solstice",
  },
  Object {
    "address": "ea92dd9a-21a3-443f-8f1c-db263192a99f",
    "creator": "0xdc76cd25977e0a5ae17155770273ad58648900d3",
    "members": Array [
      "0xdc76cd25977e0a5ae17155770273ad58648900d3",
    ],
    "motd": "Amusing universe",
    "name": "Urania",
  },
]
`;
